pmlc/target/intel_gen/subgroup_broadcast.cc

```diff
@@ -97,16 +100,56 @@ class DevectorizeImpl {
     // TODO: add additional requirements like mem alignment
     if (useBlockOps && !invalidMemScope &&
         isBlockOpTypeSupported<vector::TransferReadOp>(op)) {
-      auto newBlockReadOp =
-          builder.create<dialect::stdx::SubgroupBlockReadINTELOp>(
-              op.getLoc(), op.memref(), idxs);
-      devectorizeVectorOp(newBlockReadOp.getOperation());
-      op.replaceAllUsesWith(newBlockReadOp.getResult());
+      // Case1: vector size is the multiplication of subgroup size, in this case
+      // we use vector of size divided by subgroup size as output from
+      // SubgroupBlockReadINTELOp
+      if (op.getVectorType().getDimSize(0) > vectorSize) {
+        // Change vector dimension, we already checked earlier that dimsize %
+        // vecsize == 0
+        auto newVectorSize = op.getVectorType().getDimSize(0) / vectorSize;
+        SmallVector<int64_t, 1> newShape;
+        newShape.push_back(newVectorSize);
+        auto newMemrefType =
+            VectorType::get(newShape, op.getVectorType().getElementType());
+        auto newBlockReadOp =
+            builder.create<dialect::stdx::SubgroupBlockReadINTELOp>(
+                op.getLoc(), newMemrefType, op.memref(), idxs);
+
+        IVLOG(3, "Load Op: " << debugString(*newBlockReadOp));
+        op.replaceAllUsesWith(newBlockReadOp.getResult());
+        // Case2: output from SubgroupBlockReadINTELOp is scalar,
+        // In this case it is also needed to devectorize the operands afterwards
+      } else {
+        auto newBlockReadOp =
+            builder.create<dialect::stdx::SubgroupBlockReadINTELOp>(
+                op.getLoc(), op.memref(), idxs);
+        devectorizeVectorOp(newBlockReadOp.getOperation());
+        op.replaceAllUsesWith(newBlockReadOp.getResult());
+        IVLOG(3, "Load Op: " << debugString(*newBlockReadOp));
+      }
+      // Case3: Used buffer was allocated inside the kernel, no block reads
+      // allowed. Check if element type of the allocated memory is vector, if so
+      // then it is needed to use vector TransferReadOp for writing
+    } else if (op.memref().getDefiningOp() &&
+               dyn_cast<AllocOp>(op.memref().getDefiningOp())
```

Comment:
Don't use `dyn_cast` if the result is not being used, instead this should be `isa`

Reply:
Better yet, use a matcher :)

Reply:
Thanks Frank, actually in this case I need the dyn_cast to get Alloc's element type. But there are some other places that could be moved to matcher. I am working on adding vectorization to write ops, and will include these changes in next PR.